File: src/extern/java/BouncyCastleCryptoMac/HMac.java
package BouncyCastleCryptoMac;

import Digests.HMAC_ALGORITHM;
import Utils.AlgorithmNotSupportedException;
import Utils.Util;
import dafny.DafnySequence;

import java.math.BigInteger;

public class HMac extends _ExternBase_HMac {

  private org.bouncycastle.crypto.macs.HMac bcHMac;

  public HMac(Digests.HMAC_ALGORITHM algorithm) {
    org.bouncycastle.crypto.Digest digest;
    if(algorithm.is_HmacSHA256()) {
      digest = new org.bouncycastle.crypto.digests.SHA256Digest();
      bcHMac = new org.bouncycastle.crypto.macs.HMac(digest);
    } else if(algorithm.is_HmacSHA384()) {
      digest = new org.bouncycastle.crypto.digests.SHA384Digest();
      bcHMac = new org.bouncycastle.crypto.macs.HMac(digest);
    } else {
      throw new AlgorithmNotSupportedException(algorithm.toString() + " not supported.");
    }
  }

  @Override
  public DafnySequence<Character> getAlgorithmName() {
    return DafnySequence.asString(bcHMac.getAlgorithmName());
  }

  @Override
  public BigInteger getMacSize() {
    return BigInteger.valueOf(bcHMac.getMacSize());
  }

  @Override
  public void init(CipherParameters ps) {
    if(ps.is_KeyParameter()) {
      org.bouncycastle.crypto.params.KeyParameter bcKeyParameter = new org.bouncycastle.crypto.params.KeyParameter(ps.key);
      bcHMac.init(bcKeyParameter);
    }
  }

  @Override
  public void reset() {
    bcHMac.reset();
  }

  @Override
  public void updateSingle(byte input) {
    bcHMac.update(input);
  }

  @Override
  public void update(byte[] input , BigInteger inOff, BigInteger len) {
    bcHMac.update(input, Util.bigIntegerToInt(inOff), Util.bigIntegerToInt(len));
  }

  @Override
  public BigInteger doFinal(byte[] output, BigInteger outOff) {
    byte[] bytes = new byte[output.length - Util.bigIntegerToInt(outOff)];
    BigInteger ans = BigInteger.valueOf(bcHMac.doFinal(bytes, 0));
    System.arraycopy(bytes, 0, output, outOff.intValue(), bytes.length);
    return ans;
  }

  @Override
  public HMAC_ALGORITHM getUnderlyingDigest() {
    // TODO
    throw new UnsupportedOperationException();
  }
}

File: src/extern/dotnet/HMAC.cs
using System;
using System.Numerics;

using byteseq = Dafny.Sequence<byte>;

namespace HMAC {

    public class UnsupportedKeyDerivationAlgorithException : Exception
    {
        public UnsupportedKeyDerivationAlgorithException(KeyDerivationAlgorithms.KeyDerivationAlgorithm algorithm)
            : base(String.Format("Invalid Key Derivation Algorithm: {0}", algorithm.ToString()))
        {
        }
    }

    public partial class HMac {

        private Org.BouncyCastle.Crypto.Macs.HMac hmac;

        public HMac(KeyDerivationAlgorithms.KeyDerivationAlgorithm algorithm) {
            Org.BouncyCastle.Crypto.IDigest digest;
            if(algorithm.is_HKDF__WITH__SHA__256) {
                digest = new Org.BouncyCastle.Crypto.Digests.Sha256Digest();
            } else if(algorithm.is_HKDF__WITH__SHA__384) {
                digest = new Org.BouncyCastle.Crypto.Digests.Sha384Digest();
            } else {
                throw new UnsupportedKeyDerivationAlgorithException(algorithm);
            }
            hmac = new Org.BouncyCastle.Crypto.Macs.HMac(digest);
        }

        public void Init(CipherParameters ps) {
            if(ps.is_KeyParameter) {
                // KeyParameter/ Init should not mutate ps, but this is safer than using ps.key.Elements directly
                byte[] elemCopy = (byte[]) ps.key.Elements.Clone();
                var keyParams = new Org.BouncyCastle.Crypto.Parameters.KeyParameter(elemCopy);
                hmac.Init(keyParams);
            }
        }

        public void Update(byte input) {
            hmac.Update(input);
        }

        public void BlockUpdate(byteseq input , int inOff, int len) {
            // BlockUpdate should not mutate input, but this is safer than using input.Elements directly
            byte[] elemCopy = (byte[]) input.Elements.Clone();
            hmac.BlockUpdate(elemCopy, inOff, len);
        }

        public byteseq GetResult() {
            byte[] output = new byte[hmac.GetMacSize()];
            hmac.DoFinal(output, 0);
            return byteseq.FromArray(output);
        }
    }
}
